tree/tree/src/TTree.cxx

```diff
@@ -4457,96 +4460,121 @@ Int_t TTree::Fill()
    }
 #endif
 
-   if (fBranchRef) {
+   if (fBranchRef)
       fBranchRef->Fill();
-   }
+
    ++fEntries;
-   if (fEntries > fMaxEntries) {
+
+   if (fEntries > fMaxEntries)
       KeepCircular();
-   }
-   if (gDebug > 0) Info("TTree::Fill", " - A:  %d %lld %lld %lld %lld %lld %lld \n",
-       nbytes, fEntries, fAutoFlush,fAutoSave,GetZipBytes(),fFlushedBytes,fSavedBytes);
+
+   if (gDebug > 0)
+      Info("TTree::Fill", " - A: %d %lld %lld %lld %lld %lld %lld \n", nbytes, fEntries, fAutoFlush, fAutoSave,
+           GetZipBytes(), fFlushedBytes, fSavedBytes);
+
+   bool autoSave = false;
+   bool autoFlush = false;
 
    if (fAutoFlush != 0 || fAutoSave != 0) {
-      // Is it time to flush or autosave baskets?
       if (fFlushedBytes == 0) {
+         // If fFlushedBytes == 0, it means we never flushed or saved, so
+         // we need to check if it's time to do it and recompute the values
+         // of fAutoFlush and fAutoSave in terms of the number of entries.
          // Decision can be based initially either on the number of bytes
          // or the number of entries written.
+
          Long64_t zipBytes = GetZipBytes();
-         if ((fAutoFlush<0 && zipBytes > -fAutoFlush)  ||
-             (fAutoSave <0 && zipBytes > -fAutoSave )  ||
-             (fAutoFlush>0 && fEntries%TMath::Max((Long64_t)1,fAutoFlush) == 0) ||
-             (fAutoSave >0 && fEntries%TMath::Max((Long64_t)1,fAutoSave)  == 0) ) {
 
-            //First call FlushBasket to make sure that fTotBytes is up to date.
+         autoSave = fAutoSave < 0 ? (zipBytes > -fAutoSave) : fEntries == fAutoSave;
+         autoFlush = fAutoFlush < 0 ? (zipBytes > -fAutoFlush) : fEntries == fAutoFlush;
+
+         if (autoFlush || autoSave) {
+            // Flush baskets to update sizes in bytes, and then
+            // recompute fAutoFlush in terms of number of entries
+
             FlushBaskets();
-            OptimizeBaskets(GetTotBytes(),1,"");
-            if (gDebug > 0) Info("TTree::Fill","OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",fEntries,GetZipBytes(),fFlushedBytes);
+
+            autoFlush = false;
+            fAutoFlush = fEntries;
             fFlushedBytes = GetZipBytes();
-            fAutoFlush    = fEntries;  // Use test on entries rather than bytes
+            Long64_t totBytes = GetTotBytes();
+
+            // Optimize baskets without printing a report (option = "" below)
+            OptimizeBaskets(totBytes, /* minimum compression factor */ 1.0f, "");
+
+            if (gDebug > 0)
+               Info("TTree::Fill", "OptimizeBaskets called at entry %lld, fZipBytes=%lld, fFlushedBytes=%lld\n",
+                    fEntries, GetZipBytes(), fFlushedBytes);
```

Comment:
Due to the move ahead of the fFlushedBytes assigment, this print out now twice GetZipBytes rather than GetZipBytes and the number of bytes flushed before this point.

Reply:
Ah, yeah, I missed that, thanks. I'll move it back to the old place.